NPE from execute_cql3_query

This is due to https://github.com/palantir/atlasdb/blob/develop/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/ProfilingCassandraClient.java#L192.. if the result happens to be null, introduced by #3117.

```
Caused by: java.lang.NullPointerException: <nil>
	at com.palantir.atlasdb.keyvalue.cassandra.ProfilingCassandraClient.lambda$execute_cql3_query$156(ProfilingCassandraClient.java:192)
	at com.palantir.atlasdb.logging.KvsProfilingLogger$Monitor.lambda$log$21(KvsProfilingLogger.java:176)
	at com.palantir.atlasdb.logging.KvsProfilingLogger$Monitor.log(KvsProfilingLogger.java:187)
	at com.palantir.atlasdb.logging.KvsProfilingLogger.maybeLog(KvsProfilingLogger.java:126)
	at com.palantir.atlasdb.keyvalue.cassandra.ProfilingCassandraClient.execute_cql3_query(ProfilingCassandraClient.java:187)
	at com.palantir.atlasdb.keyvalue.cassandra.TracingCassandraClient.execute_cql3_query(TracingCassandraClient.java:144)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.palantir.tritium.proxy.InvocationEventProxy.execute(InvocationEventProxy.java:140)
	at com.palantir.tritium.proxy.InvocationEventProxy.instrumentInvocation(InvocationEventProxy.java:117)
	at com.palantir.tritium.proxy.InvocationEventProxy.handleInvocation(InvocationEventProxy.java:97)
	at com.google.common.reflect.AbstractInvocationHandler.invoke(AbstractInvocationHandler.java:86)
	at com.sun.proxy.$Proxy50.execute_cql3_query(Unknown Source)
	at com.palantir.atlasdb.keyvalue.cassandra.AutoDelegate_CassandraClient.execute_cql3_query(AutoDelegate_CassandraClient.java:171)
	at com.palantir.atlasdb.keyvalue.cassandra.qos.QosCassandraClient.lambda$execute_cql3_query$162(QosCassandraClient.java:123)
	at com.palantir.atlasdb.qos.client.AtlasDbQosClient.execute(AtlasDbQosClient.java:89)
	at com.palantir.atlasdb.qos.client.AtlasDbQosClient.executeRead(AtlasDbQosClient.java:57)
	at com.palantir.atlasdb.keyvalue.cassandra.qos.QosCassandraClient.execute_cql3_query(QosCassandraClient.java:122)
	at com.palantir.atlasdb.keyvalue.cassandra.SchemaMutationLockTables.lambda$createTableWithCustomId$98(SchemaMutationLockTables.java:111)
	at com.palantir.atlasdb.keyvalue.cassandra.CassandraClientPoolingContainer.runWithGoodResource(CassandraClientPoolingContainer.java:127)
	at com.palantir.atlasdb.keyvalue.cassandra.CassandraClientPoolingContainer.runWithPooledResource(CassandraClientPoolingContainer.java:103)
	at com.palantir.atlasdb.keyvalue.cassandra.CassandraClientPoolImpl.runWithPooledResourceRecordingMetrics(CassandraClientPoolImpl.java:429)
	at com.palantir.atlasdb.keyvalue.cassandra.CassandraClientPoolImpl.runWithRetryOnHost(CassandraClientPoolImpl.java:384)
	at com.palantir.atlasdb.keyvalue.cassandra.CassandraClientPoolImpl.runWithRetry(CassandraClientPoolImpl.java:367)
	at com.palantir.atlasdb.keyvalue.cassandra.SchemaMutationLockTables.createTableWithCustomId(SchemaMutationLockTables.java:108)
	at com.palantir.atlasdb.keyvalue.cassandra.SchemaMutationLockTables.createLockTable(SchemaMutationLockTables.java:74)
	at com.palantir.atlasdb.keyvalue.cassandra.UniqueSchemaMutationLockTable.ensureLockTableExists(UniqueSchemaMutationLockTable.java:77)
	at com.palantir.atlasdb.keyvalue.cassandra.UniqueSchemaMutationLockTable.getOnlyTable(UniqueSchemaMutationLockTable.java:48)
	at com.palantir.atlasdb.keyvalue.cassandra.CassandraKeyValueServiceImpl.tryInitialize(CassandraKeyValueServiceImpl.java:362)
	at com.palantir.atlasdb.keyvalue.cassandra.CassandraKeyValueServiceImpl.access$000(CassandraKeyValueServiceImpl.java:149)
	at com.palantir.atlasdb.keyvalue.cassandra.CassandraKeyValueServiceImpl$InitializingWrapper.tryInitialize(CassandraKeyValueServiceImpl.java:160)
	at com.palantir.async.initializer.AsyncInitializer.tryInitializeInternal(AsyncInitializer.java:147)
	at com.palantir.async.initializer.AsyncInitializer.initialize(AsyncInitializer.java:59)
	at com.palantir.atlasdb.keyvalue.cassandra.CassandraKeyValueServiceImpl.create(CassandraKeyValueServiceImpl.java:308)
	at com.palantir.atlasdb.keyvalue.cassandra.CassandraKeyValueServiceImpl.create(CassandraKeyValueServiceImpl.java:283)
	at com.palantir.atlasdb.keyvalue.cassandra.CassandraKeyValueServiceImpl.create(CassandraKeyValueServiceImpl.java:251)
	at com.palantir.atlasdb.cassandra.CassandraAtlasDbFactory.createRawKeyValueService(CassandraAtlasDbFactory.java:63)
	at com.palantir.atlasdb.factory.ServiceDiscoveringAtlasSupplier.lambda$new$14(ServiceDiscoveringAtlasSupplier.java:113)
	at com.google.common.base.Suppliers$NonSerializableMemoizingSupplier.get(Suppliers.java:164)
	at com.palantir.atlasdb.factory.ServiceDiscoveringAtlasSupplier.getKeyValueService(ServiceDiscoveringAtlasSupplier.java:126)
	at com.palantir.atlasdb.factory.ServiceDiscoveringAtlasSupplier.lambda$new$16(ServiceDiscoveringAtlasSupplier.java:122)
....
```

Full stack-trace is not posted due to internal products.

--- a/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/ProfilingCassandraClient.java
+++ b/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/ProfilingCassandraClient.java
@@ -188,8 +188,15 @@ public class ProfilingCassandraClient implements AutoDelegate_CassandraClient {
                 (KvsProfilingLogger.CallableCheckedException<CqlResult, TException>)
                         () -> client.execute_cql3_query(cqlQuery, compression, consistency),
                 (logger, timer) -> cqlQuery.logSlowResult(logger, timer),
-                (logger, cqlResult) -> logger.log("and returned {} rows, at time {}",
-                        SafeArg.of("numRows", cqlResult.getRows().size()),
-                        LoggingArgs.startTimeMillis(startTime)));
+                (logger, cqlResult) -> {
+                    if (cqlResult == null || cqlResult.getRows() == null) {
+                        logger.log("and returned null or no rows. The query was started at time {}",
+                                LoggingArgs.startTimeMillis(startTime));
+                    } else {
+                        logger.log("and returned {} rows. The query was started at time {}",
+                                SafeArg.of("numRows", cqlResult.getRows().size()),
+                                LoggingArgs.startTimeMillis(startTime));
+                    }
+                });
     }
 }

--- a/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/ProfilingCassandraClient.java
+++ b/atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/ProfilingCassandraClient.java
@@ -189,8 +189,9 @@ public class ProfilingCassandraClient implements AutoDelegate_CassandraClient {
                         () -> client.execute_cql3_query(cqlQuery, compression, consistency),
                 (logger, timer) -> cqlQuery.logSlowResult(logger, timer),
                 (logger, cqlResult) -> {
-                    if (cqlResult == null || cqlResult.getRows() == null) {
-                        logger.log("and returned null or no rows. The query was started at time {}",
+                    if (cqlResult.getRows() == null) {
+                        // different from an empty list
+                        logger.log("and returned null rows. The query was started at time {}",
                                 LoggingArgs.startTimeMillis(startTime));
                     } else {
                         logger.log("and returned {} rows. The query was started at time {}",

--- a/atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/ProfilingCassandraClientTest.java
+++ b/atlasdb-cassandra/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/ProfilingCassandraClientTest.java
@@ -1,0 +1,122 @@
+/*
+ * Copyright 2018 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.keyvalue.cassandra;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Matchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+import static org.mockito.Mockito.when;
+
+import java.nio.ByteBuffer;
+import java.util.List;
+
+import org.apache.cassandra.thrift.Column;
+import org.apache.cassandra.thrift.ColumnOrSuperColumn;
+import org.apache.cassandra.thrift.Compression;
+import org.apache.cassandra.thrift.ConsistencyLevel;
+import org.apache.cassandra.thrift.CqlResult;
+import org.apache.cassandra.thrift.CqlResultType;
+import org.apache.cassandra.thrift.CqlRow;
+import org.apache.cassandra.thrift.SlicePredicate;
+import org.apache.thrift.TException;
+import org.junit.After;
+import org.junit.Test;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.palantir.atlasdb.encoding.PtBytes;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+
+public class ProfilingCassandraClientTest {
+    private static final CqlQuery CQL_QUERY = new CqlQuery("SELECT * FROM atlasdb.foo LIMIT 1;");
+
+    private final CassandraClient delegate = mock(CassandraClient.class);
+    private final CassandraClient profilingClient = new ProfilingCassandraClient(delegate);
+
+    @After
+    public void noMoreInteractions() {
+        verifyNoMoreInteractions(delegate);
+    }
+
+    @Test
+    public void passesCqlQueriesAndResultsToAndFromDelegate() throws TException {
+        ByteBuffer byteBuffer = ByteBuffer.allocate(1);
+        CqlResult standardCqlResult = new CqlResult(CqlResultType.ROWS)
+                .setRows(ImmutableList.of(new CqlRow(byteBuffer, ImmutableList.of(new Column(byteBuffer)))));
+        setDelegateResponseToCqlQuery(standardCqlResult);
+
+        assertThat(executeCqlQuery()).isEqualTo(standardCqlResult);
+
+        verifyCqlQueryWasExecuted();
+    }
+
+    @Test
+    public void handlesNullCqlResponseFromDelegate() throws TException {
+        setDelegateResponseToCqlQuery(null);
+
+        assertThat(executeCqlQuery()).isNull();
+
+        verifyCqlQueryWasExecuted();
+    }
+
+    @Test
+    public void handlesCqlResponseWithVoidTypeFromDelegate() throws TException {
+        setDelegateResponseToCqlQuery(new CqlResult(CqlResultType.VOID));
+
+        assertThat(executeCqlQuery()).isEqualTo(new CqlResult(CqlResultType.VOID));
+
+        verifyCqlQueryWasExecuted();
+    }
+
+    @Test
+    public void handlesNullMultigetSliceResponseFromDelegate() throws TException {
+        ByteBuffer byteBuffer = ByteBuffer.wrap(PtBytes.toBytes("foo"));
+        List<ColumnOrSuperColumn> columns = ImmutableList.of(
+                new ColumnOrSuperColumn().setColumn(new Column(byteBuffer)));
+        ImmutableMap<ByteBuffer, List<ColumnOrSuperColumn>> resultMap = ImmutableMap.of(byteBuffer, columns);
+
+        when(delegate.multiget_slice(any(), any(), any(), any(), any())).thenReturn(resultMap);
+
+        assertThat(delegate.multiget_slice(
+                "getRows",
+                TableReference.createFromFullyQualifiedName("a.b"),
+                ImmutableList.of(byteBuffer),
+                new SlicePredicate(),
+                ConsistencyLevel.QUORUM)).isEqualTo(resultMap);
+
+        verify(delegate).multiget_slice(
+                "getRows",
+                TableReference.createFromFullyQualifiedName("a.b"),
+                ImmutableList.of(byteBuffer),
+                new SlicePredicate(),
+                ConsistencyLevel.QUORUM);
+    }
+
+    private void setDelegateResponseToCqlQuery(CqlResult result) throws TException {
+        when(delegate.execute_cql3_query(CQL_QUERY, Compression.NONE, ConsistencyLevel.QUORUM)).thenReturn(result);
+    }
+
+    private CqlResult executeCqlQuery() throws TException {
+        return profilingClient.execute_cql3_query(CQL_QUERY, Compression.NONE, ConsistencyLevel.QUORUM);
+    }
+
+    private void verifyCqlQueryWasExecuted() throws TException {
+        verify(delegate).execute_cql3_query(CQL_QUERY, Compression.NONE, ConsistencyLevel.QUORUM);
+    }
+}

--- a/atlasdb-client/src/main/java/com/palantir/atlasdb/logging/KvsProfilingLogger.java
+++ b/atlasdb-client/src/main/java/com/palantir/atlasdb/logging/KvsProfilingLogger.java
@@ -107,13 +107,19 @@ public class KvsProfilingLogger {
         return maybeLog(action, primaryLogger, ((loggingFunction, result) -> {}));
     }
 
+    /**
+     * Runs an action (which is a CallableCheckedException) through a {@link Monitor}, registering results on
+     * successful operations and exceptions on unsuccessful operations, as well as logging operations.
+     *
+     * Please see the documentation of {@link Monitor} for more information on how the logging functions are invoked.
+     */
     public static <T, E extends Exception> T maybeLog(CallableCheckedException<T, E> action,
             BiConsumer<LoggingFunction, Stopwatch> primaryLogger,
-            BiConsumer<LoggingFunction, T> additonalLoggerWithAccessToResult) throws E {
+            BiConsumer<LoggingFunction, T> additionalLoggerWithAccessToResult) throws E {
         if (log.isTraceEnabled() || slowlogger.isWarnEnabled()) {
             Monitor<T> monitor = Monitor.createMonitor(
                     primaryLogger,
-                    additonalLoggerWithAccessToResult,
+                    additionalLoggerWithAccessToResult,
                     slowLogPredicate);
             try {
                 T res = action.call();
@@ -167,6 +173,16 @@ public class KvsProfilingLogger {
             this.exception = ex;
         }
 
+        /**
+         * Submits the outcome of an operation to log processors, accumulating their results into a single downstream
+         * logger call. Specifically:
+         *
+         * - the primary logger (which can use a stopwatch) is always invoked
+         * - if a non-null result has been registered, call the additionalLoggerWithAccessToResult on it
+         * - if an exception has been thrown, log it
+         *
+         * The additionalLoggerWithAccessToResult may assume that its argument is non-null.
+         */
         void log() {
             stopwatch.stop();
             Consumer<LoggingFunction> logger = (loggingMethod) -> {

--- a/docs/source/release_notes/release-notes.rst
+++ b/docs/source/release_notes/release-notes.rst
@@ -54,6 +54,11 @@ develop
          - Timelock will now have more debugging info if the paxos directories fail to be created on startup.
            (`Pull Request <https://github.com/palantir/atlasdb/pull/3156>`__)
 
+    *    - |fixed|
+         - The (Thrift-backed) ``CassandraKeyValueService`` now returns correctly for CQL queries that return null.
+           Previously, they would throw an exception when we attempted to log information about the response.
+           (`Pull Request <https://github.com/palantir/atlasdb/pull/3158>`__)
+
     *    - |improved|
          - If we make a successful request to a Cassandra client, we now remove it from the Cassandra client pool's blacklist.
            Previously, removal from the blacklist would only occur after a background thread successfully refreshed the pool, meaning that requests may become stuck if Cassandra was rolling restarted.

